### ISSUE TITLE
DelegateProxy for UIApplicationDelegate?

### ISSUE THREAD

**Issue by anddygon** [ReactiveX/RxSwift#1153]
```swift
import RxCocoa
import RxSwift

class RxUIApplicationDelegateProxy: DelegateProxy, UIApplicationDelegate, DelegateProxyType {
    
    static func setCurrentDelegate(_ delegate: AnyObject?, toObject object: AnyObject) {
        let app: UIApplication = object as! UIApplication
        app.delegate = delegate as? UIApplicationDelegate
    }
    
    static func currentDelegateFor(_ object: AnyObject) -> AnyObject? {
        let app: UIApplication = object as! UIApplication
        return app.delegate
    }
    
}

extension Reactive where Base: UIApplication {
    
    var delegate: DelegateProxy {
        return RxUIApplicationDelegateProxy.proxyForObject(base)
    }
    
    var willResignActive: Observable<Void> {
        return delegate
            .methodInvoked(#selector(UIApplicationDelegate.applicationWillResignActive(_:)))
            .map{ _ in }
    }
    
}
```
I use it, like this:
```swift
UIApplication.shared.rx
            .willResignActive
            .bindNext {
                print("rx will resign active")
            }
            .disposed(by: bag)
```
But, It's doesn't work? please help me, thks advance.

**Comment by anddygon** [comment]
Fixed.  http://stackoverflow.com/questions/35575305/transform-uiapplicationdelegate-methods-into-rxswift-observables

**Comment by arctouch-augustoramos** [comment]
Could you figure it out a way to make this work?

**Comment by kodelit** [comment]
Yes, it won't work, here you have an example how it should be done. With an example of handling of `application(_:didRegister:)` method which works for me.

```
import RxSwift
import RxCocoa

extension UIApplication: HasDelegate {
    public typealias Delegate = UIApplicationDelegate
}

public class RxUIApplicationDelegateProxy: DelegateProxy<UIApplication, UIApplicationDelegate>, DelegateProxyType, UIApplicationDelegate {
    
    public init(application: UIApplication) {
        super.init(parentObject: application, delegateProxy: RxUIApplicationDelegateProxy.self)
    }
    
    public static func registerKnownImplementations() {
        self.register { RxUIApplicationDelegateProxy(application: $0) }
    }
}

extension Reactive where Base: UIApplication {
    public var delegate: RxUIApplicationDelegateProxy {
        return RxUIApplicationDelegateProxy.proxy(for: base)
    }
    

    /*
     example of delegate method handling
    */
    @available(iOS, introduced: 8.0, deprecated: 10.0)
    public var didRegisterNotificationSettings: Observable<UIUserNotificationSettings> {
        if #available(iOS 10.0, *) {
            return Observable.never()
        } else {
            return delegate.methodInvoked(#selector(UIApplicationDelegate.application(_:didRegister:)))
                .map { a in
                    return try castOrThrow(UIUserNotificationSettings.self, a[1])
            }
        }
    }
}

fileprivate func castOrThrow<T>(_ resultType: T.Type, _ object: Any) throws -> T {
    guard let returnValue = object as? T else {
        throw RxCocoaError.castingError(object: object, targetType: resultType)
    }
    
    return returnValue
}

fileprivate func castOptionalOrThrow<T>(_ resultType: T.Type, _ object: Any) throws -> T? {
    if NSNull().isEqual(object) {
        return nil
    }
    
    guard let returnValue = object as? T else {
        throw RxCocoaError.castingError(object: object, targetType: resultType)
    }
    
    return returnValue
}
```
